Cleaning AppDelegate of unused methods. UserDefaults wrapper. Rewrote Attraction Manager.

FilipTrzaska opened the pull request:
I've removed unused methods and wrapped UserDefaults usage in AppDelegate. Then I have split Attraction Manager into classes of single functionality.

Commit:
Cleaning AppDelegate of unused methods. UserDefaults wrapper. Rewrote Attraction Manager.

Myrenkar commented on Czaple/Models/FirstLaunch.swift:
```diff
@@ -0,0 +1,37 @@
+class FirstLaunch {
+    static let instance = FirstLaunch()
+    
+    func checkFirstLaunch(window: UIWindow){
```
I would suggest to split the logic. You should have one class responsible for checking and setting thins in user defaults and other that decides what should be presented based on that.

Myrenkar commented on Czaple/Models/FirstLaunch.swift:
```diff
@@ -0,0 +1,37 @@
+        if UserDefaults.standard.object(forKey: "firstTime") == nil {
+            UserDefaults.standard.set(true, forKey: "firstTime")
+        }
+        UserDefaults.standard.synchronize()
```
This is not needed anymore. Its deprecated and can be removed. It is synchronized automatically.

Myrenkar commented on Czaple/Models/FirstLaunch.swift:
```diff
@@ -0,0 +1,37 @@
+            let layout = UICollectionViewFlowLayout()
+            layout.scrollDirection = .horizontal
+            let introViewController = IntroSwipingViewController(collectionViewLayout: layout)
+            UserDefaults.standard.set(false, forKey: "firstTime")
```
You are using `UserDefaults.standard` so often that you could move it to property and it would make code cleaner :)

Myrenkar commented:
Good article for Dependency Injection to start with:

https://quickbirdstudios.com/blog/swift-dependency-injection-service-locators/

Myrenkar commented:
And this for protocol oriented programming
https://tsh.io/blog/protocol-oriented-programming-swift/

FilipTrzaska committed:
Attempt at implementing protocols

Work in progress, hope to get some feedback. I split first launcher into checker and provider but used the window attern as well, want to correct that later

FilipTrzaska committed:
Next step to protocol injecting

But appDelegate is a mess now. I need some flow to navigate through app with.

Myrenkar commented on Czaple/Abstractions/TableViewController.swift:
```diff
@@ -0,0 +1,25 @@
+
+class TableViewController: UITableViewController {
+    
+    override func viewDidLoad() {
```
You don't need to override if you are not doing anything here

Myrenkar commented on Czaple/Abstractions/ViewController.swift:
```diff
@@ -0,0 +1,26 @@
+
+class ViewController: UIViewController {
+    
+    override func viewDidLoad() {
```
Same here

Myrenkar commented on Czaple/AppDelegate.swift:
```diff
@@ -10,60 +10,23 @@ import UIKit
+        window = UIWindow()
+        guard let window = window else { return true }
+        let firstTime = firstLaunch.checkFirstLaunch()
+        introProvider.showIntro(window: window, first: firstTime)
```
Intro provider should provide viewConttroller to be shown rather you should pass window instance to it. :) 
```swift 
window.rootViewController = itroProvider.viewController(first: firstTime)
```

Myrenkar commented on Czaple/Attractions/Models/AttractionProvider.swift:
```diff
@@ -0,0 +1,106 @@
+    func getAttractions() -> [Attraction]
+}
+
+final class AttractionProvider: AttractionProviderProtocol {
```
Very nice 🥇

Myrenkar commented on Czaple/Attractions/Models/BoardAlert.swift:
```diff
@@ -0,0 +1,36 @@
+        self.boardAudioPlayer = boardAudioPlayer
+    }
+    
+    func boardAlertAction(controller: UIViewController) {
```
You should not pass viewController here, rather return this alert controller in method and then present it in place when you call that method :)

Myrenkar replied:
It's because this method should only focus on creating this alert, it will be more flexible if you let present it by someone else :)

Myrenkar commented on Czaple/Attractions/Screens/AttractionList.swift:
```diff
@@ -9,13 +9,20 @@
+    private let attractionProvider: AttractionProviderProtocol
+    private let boardAlert: BoardAlerting
+    private let cellId = "attractionCell"
+    private var locManager = CLLocationManager()
```
`locManager` can also be injected

Myrenkar commented on Czaple/Attractions/Screens/AttractionList.swift:
```diff
@@ -30,15 +37,24 @@ class AttractionList: UITableViewController {
+    // MARK: - Navigation
 
-    // MARK: - Table view data source
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```
you can remove this code :)

Myrenkar commented on Czaple/Attractions/Screens/AttractionList.swift:
```diff
@@ -30,15 +37,24 @@ class AttractionList: UITableViewController {
+extension AttractionList {
+// MARK: - Table view data source
 
     override func numberOfSections(in tableView: UITableView) -> Int {
```
you can remove this method because it does not change default implementation ( returns 1)

Myrenkar commented on Czaple/Attractions/Screens/AttractionList.swift:
```diff
@@ -48,30 +64,21 @@ class AttractionList: UITableViewController {
-        let attractionVC = AttractionView()
-        attractionVC.attraction = attractions[indexPath.row + 1]
+        let attractionVC = AttractionView(boardAlert: boardAlert)
+        attractionVC.attraction = attractionProvider.getAttractions()[indexPath.row + 1]
```
You could create a method in AttractionView that would change those attraction rather than exposing property :)

Myrenkar commented on Czaple/Attractions/Screens/AttractionView.swift:
```diff
@@ -142,7 +148,7 @@ class AttractionView: UIViewController {
             geoButton.setTitle(NSLocalizedString("Przejdź do tablic", comment: ""), for: .normal)
         default:
-            AttractionManager.instance.boardAlert(controller: self)
+            boardAlert.boardAlertAction(controller: self)
```
If you would let `AttractionView` to present this alert it would look like

```swift
let alert = boardAlert.boardAlertAction()
navigationController?.present(alert, animated: true)
```
you can see what is happening clearly

Myrenkar commented on Czaple/CzapleMapView.swift:
```diff
@@ -9,11 +9,11 @@
     var mapView: MKMapView?
     var locationManager: CLLocationManager?
-    var attractions = AttractionManager.instance.getAttractions()
+    var attractionProvider: AttractionProviderProtocol
```
change it to let :)

Myrenkar commented on Czaple/Intro/Screens/IntroSwipingViewController.swift:
```diff
@@ -74,6 +77,16 @@ class IntroSwipingViewController: UICollectionViewController, UICollectionViewDe
+        self.boardAlert = boardAlert
+        super.init(nibName: nil, bundle: nil)
+    }
+    
```
Add `@available(*, unavailable)`

Myrenkar commented on Czaple/MainView.swift:
```diff
@@ -118,22 +119,28 @@ class MainView: UIViewController {
         self.navigationController?.pushViewController(questVC, animated: true)
     }
     
+    init(attractionProvider: AttractionProviderProtocol, boardAlert: BoardAlerting) {
```
Good job here :)

Myrenkar commented on Czaple/Models/FirstLaunch.swift:
```diff
@@ -0,0 +1,26 @@
+import UIKit
+
+protocol FirstLaunchProtocol {
+    func checkFirstLaunch() -> Bool
```
You can change it to simple 
`var isFirstLaunch: Bool { get }`

Myrenkar replied:
And add method that would set it true to use it when it would be first launch. Better testability, single responsibility :)

Myrenkar commented on Czaple/Models/FirstLaunch.swift:
```diff
@@ -0,0 +1,26 @@
+class FirstLaunch: FirstLaunchProtocol {
+    private let userDefaults = UserDefaults.standard
+    
+    func checkFirstLaunch() -> Bool {
```
And then implementation would be:

```
var isFirstLaunch: Bool {
   return userDefaults.object(forKey: "firstTime") ?? false
}
```

Myrenkar commented on Czaple/Models/IntroProvider.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    func showIntro(window: UIWindow, first: Bool) {
+        if first == true {
```
`if first {` would work also. 
Generally I'm fan of returning values rather than changing them. you could simple return proper viewController like I mentioned earlier  :)

Myrenkar commented on Czaple/Models/IntroProvider.swift:
```diff
@@ -0,0 +1,40 @@
+            let layout = UICollectionViewFlowLayout()
+            layout.scrollDirection = .horizontal
+            let introViewController = IntroSwipingViewController(attractionProvider: attractionProvider, boardAlert: boardAlert)
+            UserDefaults.standard.set(false, forKey: "firstTime")
```
Here you would use of method in `FirstLaunchProtocol`
